src/ParticleEmitter.cc

```diff
@@ -0,0 +1,565 @@
+#include "Utils.hh"
+
+using namespace sdf;
+
+/// Particle emitter type strings. These should match the data in
+/// `enum class ParticleEmitterType` located in ParticleEmitter.hh.
+const std::vector<std::string> emitterTypeStrs =
+{
+  "point",
+  "box",
+  "cylinder",
+  "ellipsoid",
+};
+
+class sdf::ParticleEmitterPrivate
+{
+  /// \brief Default constructor
+  public: ParticleEmitterPrivate() = default;
+
+  /// \brief Copy constructor
+  /// \param[in] _private ParticleEmitterPrivate to move.
+  public: explicit ParticleEmitterPrivate(
+              const ParticleEmitterPrivate &_private);
+
+  // Delete copy assignment so it is not accidentally used
+  public: ParticleEmitterPrivate &operator=(
+              const ParticleEmitterPrivate &) = delete;
+
+  /// \brief Name of the particle emitter.
+  public: std::string name = "";
+
+  /// \brief Type of the particle emitter.
+  public: ParticleEmitterType type = ParticleEmitterType::POINT;
+
+  /// \brief True if the emitting should produce particles.
+  public: bool emitting = true;
+
+  /// \brief The number of seconds the emitter is active.
+  /// A value of 0 means infinite duration.
+  public: double duration = 0;
+
+  /// \brief The number of seconds each particle will 'live' for.
+  public: double lifetime = 5;
+
+  /// \brief The number of particles per second that should be emitted.
+  public: double rate = 10;
+
+  /// \brief The amount by which to scale the particles in both x and y
+  /// direction per second.
+  public: double scaleRate = 0;
+
+  /// \brief The minimum velocity for each particle (m/s).
+  public: double minVelocity = 1;
+
+  /// \brief The maximum velocity for each particle (m/s).
+  public: double maxVelocity = 1;
+
+  /// \brief The size of the emitter where the particles are sampled.
+  public: ignition::math::Vector3d size = ignition::math::Vector3d::One;
+
+  /// \brief The size of a particle.
+  public: ignition::math::Vector3d particleSize = ignition::math::Vector3d::One;
+
+  /// \brief The starting color for all particle emitted.
+  public: ignition::math::Color colorStart = ignition::math::Color::White;
+
+  /// \brief The ending color for all particle emitted.
+  public: ignition::math::Color colorEnd = ignition::math::Color::White;
+
+  /// \brief The color range image
+  public: std::string colorRangeImage = "";
+
+  /// \brief Topic used to update particle emitter properties at runtime.
+  public: std::string topic = "";
+
+  /// \brief Pose of the emitter
+  public: ignition::math::Pose3d pose = ignition::math::Pose3d::Zero;
+
+  /// \brief Frame of the pose.
+  public: std::string poseRelativeTo = "";
+
+  /// \brief Weak pointer to model's Pose Relative-To Graph.
+  public: std::weak_ptr<const sdf::PoseRelativeToGraph> poseRelativeToGraph;
+
+  /// \brief Name of xml parent object.
+  public: std::string xmlParentName = "";
+
+  /// \brief Pointer to the emitter's material properties.
+  public: std::unique_ptr<Material> material;
```

Comment:
Using `std::optional` instead of `std::unique_ptr` here will make it easier to use `ImplPtr`. See #472.

Reply:
Updated in 300e52f655d536661797edfe801dbb41cb67b3b8